BedrockServer.cpp

```diff
@@ -1776,22 +1781,48 @@ void BedrockServer::_control(unique_ptr<BedrockCommand>& command) {
             _maxConflictRetries.store(maxConflictRetries);
         }
     } else if (SIEquals(command->request.methodLine, "BlockWrites")) {
-        shared_ptr<SQLitePool> dbPoolCopy = _dbPool;
-        if (dbPoolCopy) {
-            SQLiteScopedHandle dbScope(*_dbPool, _dbPool->getIndex());
-            SQLite& db = dbScope.db();
-            db.exclusiveLockDB();
+        atomic<bool> locked(false);
+        lock_guard lock(__quiesceLock);
+        if (__quiesceThread) {
+            response.methodLine = "400 Already Blocked";
         } else {
-            response.methodLine = "404 DB Pool Not Found";
+            __quiesceThread = new thread([&]() {
+                shared_ptr<SQLitePool> dbPoolCopy = _dbPool;
+                if (dbPoolCopy) {
+                    SQLiteScopedHandle dbScope(*_dbPool, _dbPool->getIndex());
+                    SQLite& db = dbScope.db();
+                    db.exclusiveLockDB();
+                    locked = true;
+                    while (true) {
+                        if (__quiesceShouldUnlock) {
+                            db.exclusiveUnlockDB();
+                            __quiesceShouldUnlock = false;
+                            return;
+                        }
+
+                        // Wait 10ms for the next check.
+                        usleep(10'000);
+                    }
+                }
+            });
+
+            // Repeatedly wait 10ms for the lock until the thread indicates it's been acquired.
+            while (locked == false) {
+                usleep(10'000);
+            }
+
+            response.methodLine = "200 Blocked";
         }
     } else if (SIEquals(command->request.methodLine, "UnblockWrites")) {
-        shared_ptr<SQLitePool> dbPoolCopy = _dbPool;
-        if (dbPoolCopy) {
-            SQLiteScopedHandle dbScope(*_dbPool, _dbPool->getIndex());
-            SQLite& db = dbScope.db();
-            db.exclusiveUnlockDB();
+        lock_guard lock(__quiesceLock);
+        if (!__quiesceThread) {
+            response.methodLine = "400 Existing lock not found";
```

Comment:
For my use case, `200 Already Unblocked` or something would be better since if I'm trying to unblock, I don't really care if it was already unblocked. I just want to make sure it's not blocked from this point, and looking for a 200 response is simpler... What do you think?